net/gateway/connect.go

```diff
@@ -0,0 +1,234 @@
+package gateway
+
+import (
+	"crypto/ed25519"
+	"errors"
+	"fmt"
+	"go.sia.tech/core/net/mux"
+	"go.sia.tech/core/types"
+	"io"
+	"lukechampine.com/frand"
+	"net"
+)
+
+const (
+	currentVersion = "0.1"
+	minimumVersion = "0.1"
+	rejectResponse = "reject"
+	acceptResponse = "accept"
+	// listener / tcp connections not implemented yet
+	dummyAddr = "127.0.0.1:1"
+)
+
+type sessionHeader struct {
+	publicKey  ed25519.PublicKey
+	genesisID  types.BlockID
+	uniqueID   GatewayID
+	netAddress NetAddress
+}
+
+func Dial(conn net.Conn) (*Session, error) {
+	var g Session
+
+	frand.Read(g.uniqueID[:])
+	// TOOD: implement listener
+	g.netAddress = dummyAddr
+
+	// version handshake
+	if err := connectVersionHandshake(conn, currentVersion); err != nil {
+		return nil, err
+	}
+
+	// exchange session headers
+	publicKey, _, err := ed25519.GenerateKey(nil)
+	if err != nil {
+		return nil, err
+	}
+	ourHeader := sessionHeader{
+		publicKey:  publicKey,
+		genesisID:  types.GenesisBlock.ID(),
+		uniqueID:   g.uniqueID,
+		netAddress: g.netAddress,
+	}
+	if err := exchangeOurHeader(conn, ourHeader); err != nil {
+		return nil, err
+	}
+	// receive their header so we have their public key
+	remoteHeader, err := exchangeRemoteHeader(conn, ourHeader)
+	if err != nil {
+		return nil, err
+	}
+
+	session, err := mux.Dial(conn, remoteHeader.publicKey)
+	if err != nil {
+		return nil, err
+	}
+	g.mux = session
+
+	return &g, nil
+}
+
+func Accept(conn net.Conn) (*Session, error) {
+	var g Session
+
+	frand.Read(g.uniqueID[:])
+	// TOOD: implement listener
+	g.netAddress = dummyAddr
+
+	// version handshake
+	if err := acceptVersionHandshake(conn, currentVersion); err != nil {
+		return nil, err
+	}
+
+	// exchange session headers
+	publicKey, privateKey, err := ed25519.GenerateKey(nil)
+	if err != nil {
+		return nil, err
+	}
+	ourHeader := sessionHeader{
+		publicKey:  publicKey,
+		genesisID:  types.GenesisBlock.ID(),
+		uniqueID:   g.uniqueID,
+		netAddress: g.netAddress,
+	}
+	if _, err := exchangeRemoteHeader(conn, ourHeader); err != nil {
+		return nil, err
+	}
+	if err := exchangeOurHeader(conn, ourHeader); err != nil {
+		return nil, err
+	}
```

Comment:
idk if it's necessary for these to all be separate functions like they were in the old codebase. Let's try inlining them all.